includes/core/server/class-request.php

```diff
@@ -0,0 +1,306 @@
+<?php
+/**
+ * Class file for WCPay\Core\Server\Request.
+ *
+ * @package WooCommerce Payments
+ */
+
+namespace WCPay\Core\Server;
+
+use Exception;
+
+/**
+ * Base for requests to the WCPay server.
+ */
+abstract class Request {
+	/**
+	 * Contains a set of params, which the class considers immutable by others.
+	 *
+	 * Overwrite this in your class for individual properties.
+	 *
+	 * @var string[]
+	 */
+	const IMMUTABLE_PARAMS = [];
+
+	/**
+	 * Holds the parameters of the request.
+	 *
+	 * @var mixed[]
+	 */
+	private $params = [];
+
+	/**
+	 * True when `->apply_filters()` is called to protect read-only props.
+	 *
+	 * In protected mode, if somebody tries to change an immutable param,
+	 * as declared in `IMMUTABLE_PARAMS`, an exception will be thrown.
+	 *
+	 * This way important params can be safe from modifications by extensions.
+	 *
+	 * @var bool
+	 */
+	private $protected_mode = false;
+
+	/**
+	 * Creates a new instance of the class.
+	 *
+	 * @return static
+	 */
+	public static function create() {
+		return new static();
+	}
+
+	/**
+	 * Prevents the class from being constructed directly.
+	 */
+	protected function __construct() {
+		// Nothing to do here yet.
+	}
+
+	/**
+	 * Returns the needed API.
+	 *
+	 * @return string Check WCPay\Core\Server\APIs.
+	 */
+	abstract public function get_api(): string;
+
+	/**
+	 * Returns the method of the request.
+	 *
+	 * @return string See the constants in WordPress's `Requests` class.
+	 */
+	abstract public function get_method(): string;
+
+	/**
+	 * This is a legacy method, and is the same throughout the codebase.
+	 * Might be worth removing while refactoring to use the Core\Server API.
+	 *
+	 * @return bool
+	 */
+	public function is_site_specific(): bool {
+		return true;
+	}
+
+	/**
+	 * If true, the request will be signed with the user token rather than blog token. Defaults to false.
+	 *
+	 * @return bool
+	 */
+	public function should_use_user_token(): bool {
+		return false;
+	}
+
+	/**
+	 * Indicates if the raw response should be returned.
```

Comment:
This is only used for files, and once/if we move response processing out of the API client, this can be removed. That's why it's not available in the generic request class.